tests/unit/admin/views/test_classifiers.py

```diff
@@ -1,134 +0,0 @@
-        l2, l3, l4, l5 = parent_levels
-        parent = ClassifierFactory(l2=l2, l3=l3, l4=l4, l5=l5, classifier="Parent")
-
-        db_request.params = {"parent_id": parent.id, "child": "Foobar"}
-        db_request.session.flash = pretend.call_recorder(lambda *a, **kw: None)
-        db_request.route_path = lambda *a: "/the/path"
-
-        views.AddClassifier(db_request).add_child_classifier()
-
-        new = (
-            db_request.db.query(Classifier)
-            .filter(Classifier.classifier == "Parent :: Foobar")
-            .one()
-        )
-
-        new_l2, new_l3, new_l4, new_l5 = expected_levels
-        assert new.l2 == new_l2 if new_l2 is not None else new.id
-        assert new.l3 == new_l3 if new_l3 is not None else new.id
-        assert new.l4 == new_l4 if new_l4 is not None else new.id
-        assert new.l5 == new_l5 if new_l5 is not None else new.id
-
-    def test_add_parent_classifier(self, db_request):
-        db_request.params = {"parent": "Foo :: Bar"}
-        db_request.session.flash = pretend.call_recorder(lambda *a, **kw: None)
-        db_request.route_path = lambda *a: "/the/path"
-
-        views.AddClassifier(db_request).add_parent_classifier()
-
-        new = (
-            db_request.db.query(Classifier)
-            .filter(Classifier.classifier == "Foo :: Bar")
-            .one()
-        )
-
-        assert new.l2 == new.id
-        assert new.l3 == 0
-        assert new.l4 == 0
-        assert new.l5 == 0
-
-    @pytest.mark.parametrize(
-        "parent_classifier, parent_levels, expected_levels",
-        [
-            ("private", (2, 0, 0, 0), (2, None, 0, 0)),
-            ("private", (2, 3, 0, 0), (2, 3, None, 0)),
-            ("private", (2, 3, 4, 0), (2, 3, 4, None)),
-            ("Private", (2, 0, 0, 0), (2, None, 0, 0)),
-            ("Private", (2, 3, 0, 0), (2, 3, None, 0)),
-            ("Private", (2, 3, 4, 0), (2, 3, 4, None)),
-            ("PrIvAtE", (2, 0, 0, 0), (2, None, 0, 0)),
-            ("PrIvAtE", (2, 3, 0, 0), (2, 3, None, 0)),
-            ("PrIvAtE", (2, 3, 4, 0), (2, 3, 4, None)),
-        ],
-    )
-    def test_add_private_child_classifier(
-        self, db_request, parent_classifier, parent_levels, expected_levels
-    ):
-        l2, l3, l4, l5 = parent_levels
-        parent = ClassifierFactory(
-            l2=l2, l3=l3, l4=l4, l5=l5, classifier=parent_classifier
-        )
-
-        db_request.params = {"parent_id": parent.id, "child": "Foobar"}
-        db_request.session.flash = pretend.call_recorder(lambda *a, **kw: None)
-        db_request.route_path = lambda *a: "/the/path"
-
-        with pytest.raises(sqlalchemy.exc.IntegrityError):
-            views.AddClassifier(db_request).add_child_classifier()
-
-    @pytest.mark.parametrize("parent_classifier", ["private", "Private", "PrIvAtE"])
-    def test_add_private_parent_classifier(self, db_request, parent_classifier):
```

Comment:
I liked that there were tests preventing the upload of `Private ::` classifiers.
Is it now tested elsewhere ?

Reply:
We still retain a CheckConstraint on the Classifier model to ensure that nothing that case-insensitive matches `'private ::%'` can be created in the DB. Which would act as a backstop in the event that a `trove-classifiers` release included such a classifier. It probably makes sense to do two things 1) include a test in warehouse to verify that all classifiers from a given release of `trove-classifiers` can be created which would catch such a condition in CI rather than at deploy 2) include a test in `trove-classifiers` itself.

Reply:
Great points. Done!